In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf
import keras

#11.1 그레이디언트 소실과 폭주 문제

##11.1.1 글로럿과 He 초기화

In [ ]:
keras.layers.Dense(10, activation="relu", kernel_initializer="he_normal")

In [ ]:
he_avg_init = keras.initializers.VarianceScaling(scale=2., mode='fan_avg', distribution='uniform')
keras.layers.Dense(10, activation="sigmoid", kernel_initializer=he_avg_init)

##11.1.2 수렴하지 않는 활성화 함수

In [ ]:
#LeakyReLU 활성화 함수하려면 아예 층을 만들어서 추가해야한다..

model = keras.models.Sequential([
    keras.layers.Dense(30, activation="relu", input_shape=[8]),
    keras.layers.Dense(30, activation="relu"),
    keras.layers.Dense(10, kernel_initializer="he_normal"),
    keras.layers.LeakyReLU(alpha=0.2),       #활성화함수 층 추가..
    keras.layers.Dense(1)
])

In [ ]:
#SELU 활성화함수는 그냥 매개변수 조작만..

layer = keras.layers.Dense(10, activation="selu", kernel_initializer="lecun_normal")

##11.1.3 배치 정규화

### 케라스로 배치 정규화 구현하기

In [ ]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28,28]),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(300, activation='elu', kernel_initializer="he_normal"),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(100, activation='elu', kernel_initializer="he_normal"),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(10, activation='softmax')
])

In [ ]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
batch_normalization (BatchNo (None, 784)               3136      
_________________________________________________________________
dense_7 (Dense)              (None, 300)               235500    
_________________________________________________________________
batch_normalization_1 (Batch (None, 300)               1200      
_________________________________________________________________
dense_8 (Dense)              (None, 100)               30100     
_________________________________________________________________
batch_normalization_2 (Batch (None, 100)               400       
_________________________________________________________________
dense_9 (Dense)              (None, 10)               

- 배치 정규화 층에서 입력마다 네 개의 파라미터 $\gamma, \beta, \mu, \sigma$를 학습한다..

- $\mu, \sigma$는 이동평균이고, 역전파로 학습되지 않으므로 Non-trainable 파라미터로 분류된다..
    - Non-trainable params : 2 * (784 + 300 + 100) = 2368

In [ ]:
#Non-trainable params확인..
#이동평균, 이동표준편차는 추정되는 것.. -> Non-trainable..

[(var.name, var.trainable) for var in model.layers[1].variables]

[('batch_normalization/gamma:0', True),
 ('batch_normalization/beta:0', True),
 ('batch_normalization/moving_mean:0', False),
 ('batch_normalization/moving_variance:0', False)]

In [ ]:
#제거될 속성..
model.layers[1].updates

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py:1348: UserWarning: `layer.updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  warnings.warn('`layer.updates` will be removed in a future version. '


[]

##11.1.4 그레이디언트 클리핑

#11.2 사전훈련된 층 재사용하기

##11.2.1 케라스를 사용한 전이 학습

##11.2.2 비지도 사전훈련

##11.2.3 보조 작업에서 사전훈련

#11.3 고속 옵티마이저

##11.3.1 모멘텀 최적화

##11.3.2 네스테로프 가속 경사

##11.3.3 AdaGrad

##11.3.4 RMSProp

##11.3.5 Adam과 Nadam 최적화

##11.3.6 학습률 스케쥴링

#11.4 규제를 사용해 과대적합 피하기

##11.4.1 $l_1$과 $l_2$ 규제

##11.4.2 드롭아웃

##11.4.3 몬테 카를로 드롭아웃

##11.4.4. 맥스-노름 규제

#11.5 요약 및 실욕적인 가이드라인

#11.6 연습문제